### PR TITLE
Readd Transf

### PR DESCRIPTION
**Pull request by james-d-mitchell**

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
I think this is ready to go, @Joseph-Edwards

File: dev-environment.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+channels:
+  - conda-forge
+
+dependencies:
+  - fmt
+  - pip: 
+    - accepts
+    - bs4
+    - lxml
```
Comment (Joseph-Edwards) [inline dev-environment.yml:11]: I checked conda-forge and I think we can do this in the interest of minimising conda-pip interactions
```suggestion
dependencies:
  - fmt
  - bs4
  - lxml
  - pip: 
    - accepts
```

**Comment by Joseph-Edwards** [comment]
Modulo the tests passing (which I assume they will after https://github.com/libsemigroups/libsemigroups/pull/544 is merged) and my minor suggestion above, this looks good to me. That said, I'd like to build and read the docs just to be doubly sure. I'll do this first thing tomorrow morning.

File: src/transf.cpp [hunk 2]
```diff
@@ -43,99 +32,634 @@ namespace py = pybind11;
...
+   >>> x = Perm1.one(8)
+   >>> x
+   Perm1([0, 1, 2, 3, 4, 5, 6, 7])
+   >>> x.degree()
+   8
+   >>> x.undef()
+   255
+   >>> x.swap(y)
+   >>> x, y
+   (Perm1([5, 2, 0, 1, 3, 4, 6]), Perm1([0, 1, 2, 3, 4, 5, 6, 7])
```
Comment (Joseph-Edwards) [inline src/transf.cpp:568]:
```suggestion
   (Perm1([5, 2, 0, 1, 3, 4, 6]), Perm1([0, 1, 2, 3, 4, 5, 6, 7]))
```

File: docs/source/data-structures/elements/transf-helpers.rst [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+and similarly for ``PPerm`` and ``Perm``.
+
+Full API
+--------
+
+.. py:function:: one(f: Transf | PPerm | Perm) -> Transf | PPerm | Perm
+
+   Returns the identity on the same number of points as the degree of **f**.
+   This function returns a newly constructed object of the same type as **f**
```
Comment (Joseph-Edwards) [inline docs/source/data-structures/elements/transf-helpers.rst:28]:
```suggestion
   Returns the identity on the same number of points as the degree of *f*.
   This function returns a newly constructed object of the same type as *f*
```

File: docs/source/data-structures/elements/transf-helpers.rst [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+   that fixes every value from ``0`` to ``f.degree()``.
+   
+   :param f: the transformation.
+   :type f: Transf | PPerm | Perm
+   
+   :complexity: Linear in the degree of **f**.
```
Comment (Joseph-Edwards) [inline docs/source/data-structures/elements/transf-helpers.rst:34]:
```suggestion
   :complexity: Linear in the degree of *f*.
```

File: docs/source/data-structures/elements/transf-helpers.rst [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+   
+   :returns: The identity.
+   :rtype: Transf | PPerm | Perm 
+
+.. py:function:: inverse(f: PPerm | Perm) -> PPerm | Perm
+  
+   Returns the inverse of a partial perm or permutation.
+  
+   This function returns a newly constructed inverse of ``f``. The *inverse* of
```
Comment (Joseph-Edwards) [inline docs/source/data-structures/elements/transf-helpers.rst:43]:
```suggestion
   This function returns a newly constructed inverse of *f*. The *inverse* of
```

File: docs/source/data-structures/elements/transf-helpers.rst [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+   a partial permutation ``f`` is the partial perm ``g`` such that ``fgf =
+   f`` and ``gfg = g``.
+  
+   :param f: the partial perm.
+   :type f: PPerm | Perm
+   
+   :complexity: Linear in the degree of **f**.
```
Comment (Joseph-Edwards) [inline docs/source/data-structures/elements/transf-helpers.rst:50]:
```suggestion
   :complexity: Linear in the degree of *f*.
```

File: docs/source/data-structures/elements/transf-helpers.rst [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+   :returns: The inverse of ``f``.
+   :rtype: PPerm | Perm 
+
+.. py:function:: domain(f: Transf | PPerm | Perm) -> List[int]
+  
+   Returns the set of points where a partial transformation is defined.
+   
+   Returns a ``List[int]`` containing those values ``i`` such that:
+
+   * :math:`i\in \{0, \ldots, n - 1\}` where `n` is the degree of **f**; and
```
Comment (Joseph-Edwards) [inline docs/source/data-structures/elements/transf-helpers.rst:61]:
```suggestion
   * :math:`i\in \{0, \ldots, n - 1\}` where `n` is the degree of *f*; and
```

File: docs/source/data-structures/elements/transf-helpers.rst [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+   * ``f[i] != UNDEFINED``.
+   
+   :param f: the transformation.
+   :type f: Transf | PPerm | Perm
+   
+   :complexity: Linear in the degree of **f**.
```
Comment (Joseph-Edwards) [inline docs/source/data-structures/elements/transf-helpers.rst:67]:
```suggestion
   :complexity: Linear in the degree of *f*.
```

File: docs/source/data-structures/elements/transf-helpers.rst
Comment (Joseph-Edwards) [inline docs/source/data-structures/elements/transf-helpers.rst:1]: I think we decided on the convention that function parameters to be wrapped in single pair of asterisks like `*f*`. I have made a few suggestions, but I don't think I got all of them.

**Review by Joseph-Edwards — CHANGES_REQUESTED** [review]
I've read had a more thorough read through the docs now, and it all looks very good to me. I have a few minor comments before I'd be happy to merge.
1. For consistency, should the new pages in the doc have a table of contents as in `Action` and `Knuth-Bendix`?
2. After the signatures of some of the functions, some return types have hyperlinks and some don't. For example, in `PPerm(dom: List[int], img: List[int], deg: int) → StaticPPerm16 | PPerm1 | PPerm2 | PPerm4`, `StaticPPerm16` gets a hyperlink, but the others don't. It's not immediately clear if there's a nice fix for this (or if a nice fix is even necessary), but it would be nice to do if there is.
3. If I understood correctly, @james-d-mitchell asked if it was possible to show the constructor when documenting a class. I couldn't remember how to do that at the time, but I now recall that is done with something like this:
```rst
.. autoclass:: StaticTransf16
   :class-doc-from: class
   :members:
   :inherited-members:
   :special-members: __init__
```

With the above comments considered, I would be happy to merge

**Pull request merged**